# The goal of this section is to familiarize ourselves with the data we use.  
I'll start by looking at the raw data.

## We are looking for answers to the following questions:
1. What columns are there in the file and which data type (text, number, date)?
2. How many rows does the dataset consist of?
3. Are there any missing values ​​(null / NaN) and if so, in which columns?
4. Are there completely duplicate rows?
5. What format is the date field in? (Did pandas read it as a string or recognize it as a date?)

## We will not solve the questions in order, but at the end you will find a [summary](#summary-table) where you can find all the answers.  
Get data from csv-s and manually check to readings were complete.  
Originally, I thought of reading the two csvs into a variable and then combining them with the concat() function. The problem with this is that the concat() function creates a separate copy when I join the two dataframes together. To avoid this, I read them in as a list, so the data only appears in memory once.  
The ignore_index is an important switch here, as joining the two csvs will prevent it from starting the indexing from the beginning.  
Since I looked at the raw csv files, I saw that they also contain dates. These must be explicitly provided to Pandas.

In [1]:
import pandas as pd

df = pd.concat([pd.read_csv(f, parse_dates=['trans_date_trans_time', 'dob']) for f in ['fraudTest.csv', 'fraudTrain.csv']], ignore_index=True)
df.head()

,Unnamed: 0,trans_date_trans_time,cc_num,merchant,category,amt,first,last,gender,street,...,lat,long,city_pop,job,dob,trans_num,unix_time,merch_lat,merch_long,is_fraud
0,0,2020-06-21 12:14:25,2291163933867244,fraud_Kirlin and Sons,personal_care,2.86,Jeff,Elliott,M,351 Darlene Green,...,33.9659,-80.9355,333497,Mechanical engineer,1968-03-19,2da90c7d74bd46a0caf3777415b3ebd3,1371816865,33.986391,-81.200714,0
1,1,2020-06-21 12:14:33,3573030041201292,fraud_Sporer-Keebler,personal_care,29.84,Joanne,Williams,F,3638 Marsh Union,...,40.3207,-110.4360,302,"Sales professional, IT",1990-01-17,324cc204407e99f51b0d6ca0055005e7,1371816873,39.450498,-109.960431,0
2,2,2020-06-21 12:14:53,3598215285024754,"fraud_Swaniawski, Nitzsche and Welch",health_fitness,41.28,Ashley,Lopez,F,9333 Valentine Point,...,40.6729,-73.5365,34496,"Librarian, public",1970-10-21,c81755dbbbea9d5c77f094348a7579be,1371816893,40.495810,-74.196111,0
3,3,2020-06-21 12:15:15,3591919803438423,fraud_Haley Group,misc_pos,60.05,Brian,Williams,M,32941 Krystal Mill Apt. 552,...,28.5697,-80.8191,54767,Set designer,1987-07-25,2159175b9efe66dc301f149d3d5abf8c,1371816915,28.812398,-80.883061,0
4,4,2020-06-21 12:15:17,3526826139003047,fraud_Johnston-Casper,travel,3.19,Nathan,Massey,M,5783 Evan Roads Apt. 465,...,44.2529,-85.0170,1126,Furniture designer,1955-07-06,57ff021bd3f328f8738bb535c302a31b,1371816917,44.959148,-85.884734,0


In [2]:
df.shape

(1852394, 23)

We already answered the second quoestion. Let's check the data types:

In [3]:
df.dtypes

Unnamed: 0                        int64
trans_date_trans_time    datetime64[us]
cc_num                            int64
merchant                            str
category                            str
amt                             float64
first                               str
last                                str
gender                              str
street                              str
city                                str
state                               str
zip                               int64
lat                             float64
long                            float64
city_pop                          int64
job                                 str
dob                      datetime64[us]
trans_num                           str
unix_time                         int64
merch_lat                       float64
merch_long                      float64
is_fraud                          int64
dtype: object

In [4]:
df.info(verbose=True, show_counts=True)

<class 'pandas.DataFrame'>
RangeIndex: 1852394 entries, 0 to 1852393
Data columns (total 23 columns):
 #   Column                 Non-Null Count    Dtype         
---  ------                 --------------    -----         
 0   Unnamed: 0             1852394 non-null  int64         
 1   trans_date_trans_time  1852394 non-null  datetime64[us]
 2   cc_num                 1852394 non-null  int64         
 3   merchant               1852394 non-null  str           
 4   category               1852394 non-null  str           
 5   amt                    1852394 non-null  float64       
 6   first                  1852394 non-null  str           
 7   last                   1852394 non-null  str           
 8   gender                 1852394 non-null  str           
 9   street                 1852394 non-null  str           
 10  city                   1852394 non-null  str           
 11  state                  1852394 non-null  str           
 12  zip                    1852394 non-null

Let's check for any duplicate rows.
Since Pandas looks at all the data in the row when checking for duplication and now we includes a column in our data that will not be needed (unnamed), it is worth filtering the duplications based on a subset, in this case based on the transaction numbers, as this must be unique.

In [5]:
df = df.drop_duplicates(subset=['trans_num'])
df.shape

(1852394, 23)

## Here are my answers to the predetermined questions, based on the data above:
1. What columns are there in the file and which data type (text, number, date)?
    As you can see, Pandas cleverly sorted out the data types: **float64(5), int64(6), str(10)***
2. How many rows does the dataset consist of?
    There are **1852394** rows.
3. Are there any missing values ​​(null / NaN) and if so, in which columns?
    The data set is complete and contains **no empty values**.
4. Are there completely duplicate rows?
    No, there were **no duplication** in the dataset
5. What format is the date field in? (Did pandas read it as a string or recognize it as a date?)
    We already told it which columns would be of date type when reading.

<a id="summary-table"></a>
## Summary table:
| Question / Metric | Result / Finding |
| -------- | ------- |
| Total Rows | 1,852,394 |
| Total Columns | 23 |
| Missing Values | 0 (Dataset is clean)|
| Duplicates | 0 unique transactions duplicated|
| Date Ranges | Parsed to datetime64[us] |


## We found the answers to the questions. Let’s dig a little deeper!  
The dataset we’re working with seems clean enough, so I don’t see the need to separate data cleaning from discovery. Let’s check that the columns that should contain unique values ​​are indeed unique. Also, let’s check the columns that contain categorical results. Check that they don’t contain typos or invalid values.  
As we saw earlier, the first column contains row numbers, without names. This will not be needed, as both Pandas and SQL automatically number the rows. So let's start by deleting the first column:

In [10]:
df.drop(columns=["Unnamed: 0"], inplace=True)
df.info(verbose=True)

<class 'pandas.DataFrame'>
RangeIndex: 1852394 entries, 0 to 1852393
Data columns (total 22 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   trans_date_trans_time  datetime64[us]
 1   cc_num                 int64         
 2   merchant               str           
 3   category               str           
 4   amt                    float64       
 5   first                  str           
 6   last                   str           
 7   gender                 str           
 8   street                 str           
 9   city                   str           
 10  state                  str           
 11  zip                    int64         
 12  lat                    float64       
 13  long                   float64       
 14  city_pop               int64         
 15  job                    str           
 16  dob                    datetime64[us]
 17  trans_num              str           
 18  unix_time              int64     

Let's summarize which columns we will examine for their uniqueness and which ones for their category distribution.  
Columns containing unique elements: [cc_num, trans_num]  
Categorical columns: [merchant, category, gender, city, state, is_fraud]

In [13]:
df['cc_num'].nunique()

999

In [12]:
df['trans_num'].nunique()

1852394

As you can see in the results, cc_num identified 999 customers. However, trans_num is completely unique in each row.

In [15]:
df['merchant'].value_counts()

merchant
fraud_Kilback LLC                        6262
fraud_Cormier LLC                        5246
fraud_Schumm PLC                         5195
fraud_Kuhn LLC                           5031
fraud_Boyer PLC                          4999
                                         ... 
fraud_Douglas, DuBuque and McKenzie      1101
fraud_Treutel-King                       1098
fraud_Satterfield-Lowe                   1095
fraud_Hahn, Douglas and Schowalter       1091
fraud_Ritchie, Bradtke and Stiedemann    1090
Name: count, Length: 693, dtype: int64

In [14]:
df['category'].value_counts()

category
gas_transport     188029
grocery_pos       176191
home              175460
shopping_pos      166463
kids_pets         161727
shopping_net      139322
entertainment     134118
food_dining       130729
personal_care     130085
health_fitness    122553
misc_pos          114229
misc_net           90654
grocery_net        64878
travel             57956
Name: count, dtype: int64

In [16]:
df['gender'].value_counts()

gender
F    1014749
M     837645
Name: count, dtype: int64

In [17]:
df['city'].value_counts()

city
Birmingham     8040
San Antonio    7312
Utica          7309
Phoenix        7297
Meridian       7289
               ... 
Norfolk           7
Karns City        7
Bruce             7
Streator          7
Wever             6
Name: count, Length: 906, dtype: int64

In [18]:
df['state'].value_counts()

state
TX    135269
NY    119419
PA    114173
CA     80495
OH     66627
MI     65825
IL     62212
FL     60775
AL     58521
MO     54904
MN     45433
AR     44611
NC     43134
VA     41756
WI     41738
SC     41731
KY     40981
IN     39539
IA     38804
OK     38050
MD     37345
GA     37340
WV     36529
NJ     35131
NE     34425
KS     32939
MS     30021
LA     29953
WY     27776
WA     27040
OR     26408
TN     24913
ME     23433
NM     23427
ND     21183
CO     19766
SD     17574
MA     17562
VT     16812
MT     16806
AZ     15362
UT     15357
NH     11727
CT     10979
NV      8058
ID      8035
DC      5130
HI      3649
AK      2963
RI       745
DE         9
Name: count, dtype: int64

In [19]:
df['is_fraud'].value_counts()

is_fraud
0    1842743
1       9651
Name: count, dtype: int64

The dataset contains widely distributed and well-collected data. **No further data cleaning is expected to be required.**